gitpod-io/gitpod
# [integration tests] Support Git repo fixtures

## Description
Today our integration tests rely on the undocumented, unspecified structure of some repo on GitHub. This makes our tests brittle and subject to random failures which are hard to recover from, because no one remembers how the repo should look like.

This PR extends integration tests to create Git repositories before they're needed. For example, if we needed a repo of a certain structure, we can now produce it using:
```Go
token, err := integration.GitHubToken(ctx, username, api)
repo, err := git.MaterialiseGitHubRepo(ctx, token, org, git.TempRepo(true), git.Ops{
    git.OpAddFile(".gitpod.yml", "image:\n  file: Dockerfile"),
    git.OpAddFile("Dockerfile", "FROM gitpod/workspace-full:latest"),
    git.OpCommitAll("initial commit"),
})
t.Cleanup(func() {
    repo.Delete(ctx)
})

integration.LaunchWorkspaceFromContextURL(ctx, repo.ContextURL(), username, api)
```
A concrete use example can be found [here](https://github.com/gitpod-io/gitpod/blob/4a3f80c38345aa045abcf0a097b493d23640734a/test/tests/workspace/git_test.go#L113-L131).

Note: 
- to use `integration.GitHubToken` the user needs at least the `repo` and `read_user` scopes.
- because Gitpod cannot request the `repo_delete` scope, the `repo.Delete()` operation is likely to fail. 

### Repo's from files
Sometimes repo's are too complicated to be described in code. Then, we can place them alongside the test and use `git.OpsFromFixture("some-path/")` to produce the test repo.

## How to test
1. log in on the Gitpod installation you'll run the test against
2. Ensure your user has the correct permissions
  <img width="1006" alt="image" src="https://user-images.githubusercontent.com/3210701/154670456-7013859d-6065-4e32-95d2-5d415a45bdd7.png">
3. Run the test

    cd test/tests/workspace
    go test  -run TestGitActions -v ./... -namespace=staging-main -kubeconfig=/home/gitpod/.kube/config -username=<your-github-user>

4. Check your account and look for a repo named something like `testing-1645181683-1443635317331776148`

## Release Notes

```release-note
NONE
```
